### PR TITLE
Check for CMAKE_C_COMPILER_ID not CXX

### LINKED ISSUE #55412: [Android] Building fails with Android NDK r22 on OSX?
1. Download Android NDK r22
2. Build with `export ANDROID_NDK_ROOT=/path/to/android-ndk-r22 ./build.sh --os android --arch x64 -c Release`

Expected:

Build succeeds

Actual:

```
EXEC : error : /Applications/Xcode_12.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive library: libSystem.Native.a the table of contents is empty (no object file members in the library define global symbols) [/Users/alklig/work/dotnet-runtime/runtime-bugs/src/libraries/Native/build-native.proj]
EXEC : error : /Applications/Xcode_12.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive library: libSystem.Security.Cryptography.Native.Android.a the table of contents is empty (no object file members in the library define global symbols) [/Users/alklig/work/dotnet-runtime/runtime-bugs/src/libraries/Native/build-native.proj]
EXEC : error : /Applications/Xcode_12.5.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive library: libSystem.IO.Compression.Native.a the table of contents is empty (no object file members in the library define global symbols) [/Users/alklig/work/dotnet-runtime/runtime-bugs/src/libraries/Native/build-native.proj]
```

Build fails while linking `libSystem.Native.a` with:

Note that it's **Xcode**'s linker that is complaining.  So something about the toolchain detection is not picking up the linker from the NDK and falling back to whatever it finds on the PATH.

Possibly see https://github.com/dotnet/runtime/pull/51876 which attempted to support r22 apparently did not get everything.

---

I have this in `ndk/source.properties`

```
Pkg.Desc = Android NDK
Pkg.Revision = 22.0.7026061
```

Have not tried the latest r22b  22.1.7171670 at this time.

### PR DESCRIPTION
**Pull request by directhex**
In some cases, `CMAKE_CXX_COMPILER_ID` might not be populated, which causes the toolchain prefix checks to be ignored, which causes the build to look for `ranlib` not `llvm-ranlib`, which causes it to use XCode `ranlib` instead of Android NDK `ranlib` for builds, which doesn't work.

Hopefully (we'll see what CI says) this doesn't regress other builds.

Closes: #55412

### CODE REVIEW

**Comment by jkotas** [comment]
cc @am11

File: eng/native/configuretools.cmake
Comment (am11) [inline eng/native/configuretools.cmake:18]: This should also be changed to `LANG:C`:
```suggestion
      set(TOOLSET_PREFIX "${CMAKE_C_COMPILER_TARGET}-")
```

**Review by am11 — COMMENTED** [review]
LGTM

**Comment by directhex** [comment]
Failing lane is an unrelated timeout issue, will do a separate PR to make that one not run on PR

**Comment by am11** [comment]
@jkotas, @directhex, looking at it again with fresh eyes I just remember making this ditto change in the past https://github.com/dotnet/runtime/commit/ba50840433578d97de4ad8bb9d2a2b45e7455b30 to use "correct" objcopy. Unfortunately, it caused binary size regression and ultimately we had to revert it.

The issue was that on my local system (Ubuntu) I was using llvm-10 at the time but on build systems, we were (and still are) using llvm-9 in most cases. There is a bug in llvm-9 objcopy which fails to strip the symbols. LLVM 10 onwards has that bug fixed (it was never backported).

Therefore, I think we should hold this until we update official baseline to clang 10 or above, and continue to use GNU's objcopy instead of LLVM9's one.

**Comment by am11** [comment]
@directhex, the effect was on objcopy but the change was exactly what you have `CMAKE_CXX_COMPILER_ID` to `CMAKE_C_COMPILER_ID` that broke the world.

**Comment by directhex** [comment]
@am11 I can't reproduce that.

Using mcr.microsoft.com/dotnet-buildtools/prereqs:centos-7-20210714125435-9b5bbc2, on git master:

```
-rwxr-xr-x 1 root root 14M Jan 26 16:23 /home/directhex/Projects/runtime/artifacts/bin/coreclr/Linux.x64.Debug/./libcoreclr.so
-rw-r--r-- 1 root root 79M Jan 26 16:23 /home/directhex/Projects/runtime/artifacts/bin/coreclr/Linux.x64.Debug/./libcoreclr.so.dbg
```

this branch:
```
-rwxr-xr-x 1 root root 14M Jan 26 16:29 /home/directhex/Projects/runtime/artifacts/bin/coreclr/Linux.x64.Debug/./libcoreclr.so
-rw-r--r-- 1 root root 79M Jan 26 16:29 /home/directhex/Projects/runtime/artifacts/bin/coreclr/Linux.x64.Debug/./libcoreclr.so.dbg
```

Looking at the build logs and CMake output, `CMAKE_C_COMPILER_ID` and `CMAKE_CXX_COMPILER_ID` are both set correctly to Clang on Linux, so the behaviour should not be changed by this PR.

Actually, looking even closer, the core behaviour here is:
1) The `TOOLSET_PREFIX` is set to `llvm-` both before and after my PR: https://github.com/dotnet/runtime/blob/df69293bdbe75bf5903583c472fcd257119dbda4/eng/native/configuretools.cmake#L9-L22
2) `ar`, `nm`, `ranlib`, `link` and `objdump` are set based on this value of `TOOLSET_PREFIX`: https://github.com/dotnet/runtime/blob/df69293bdbe75bf5903583c472fcd257119dbda4/eng/native/configuretools.cmake#L42-L51
3) `TOOLSET_PREFIX` is reconfigured to be `${ANDROID_TOOLCHAIN_PREFIX}` for Android, `${TOOLCHAIN}-` if cross compiling for ARM or s390x, and crucially blank otherwise - this means the value for `TOOLSET_PREFIX` is now blank on linux-x64: https://github.com/dotnet/runtime/blob/df69293bdbe75bf5903583c472fcd257119dbda4/eng/native/configuretools.cmake#L53-L60
4) set `objcopy` with this blank prefix (which means GNU objcopy will be used, due to the empty `TOOLSET_PREFIX`)

The reasoning given in #49092 for the revert is definitely the linux-x64 case

**Comment by am11** [comment]
> set objcopy with this blank prefix (which means GNU objcopy will be used, due to the empty TOOLSET_PREFIX)

Cool. Thanks for confirming it. We can update it to also use matching/correct toolchain, once baseline is updated to clang 10 or beyond.

**Comment by directhex** [comment]
Do we have a CI lane to check for size regression? Because I don't see how this change ever affected the build in the way reported in #49092

**Comment by am11** [comment]
Not sure if there is such a lane, but there is a label to that effect: https://github.com/dotnet/runtime/issues?q=is:open+is:issue+label:size-reduction.

I think in this case, the shipping installer package (which includes runtimes+sdks+templates) was manually compared with the previous release at the time. Similar inspection is done for file count differences in packages near the releases to prevent unintended files from getting shipped (or find missing essential files or even duplicate files in nuget packages). Some of those targets are automated and run only during the official builds AFAIK.

**Comment by am11** [comment]
> so the behaviour should not be changed by this PR.

Technically the behavior does change for `libs.native` subset due to second argument to `project` command: https://github.com/dotnet/runtime/blob/5a9bcde1d52b8b5a6c7b70883704bcb349c54c12/src/native/libs/CMakeLists.txt#L18
In this case, cmake only locates and activates C toolchain and `CMAKE_CXX_COMPILER_ID` is empty string, but `CMAKE_C_COMPILER_ID` has a value.

**Comment by directhex** [comment]
Okay, it seems it _does_ have an effect on libs.native - for `ar`, `link`, `nm`, `objdump` and `ranlib` it picks the LLVM versions instead of GNU

But it still uses GNU `objcopy` either way, due to TOOLSET_PREFIX being wiped, so the file size is no different on libSystem.Native.so before or after the patch

**Comment by directhex** [comment]
@am11 can I have an approval, if you're confident with my analyses of the file size concern? Is there someone else we should ask?

**Comment by am11** [comment]
LGTM, thank you!

cc @janvorli

**Review by janvorli — APPROVED** [review]
LGTM, thank you!

**Pull request merged**
